### PR TITLE
[Private media files] Use API proxy for private atomic sites

### PR DESCRIPTION
**Pull request by jblz**
#### Changes proposed in this Pull Request

This PR is related to [private atomic sites project](https://wp.me/paObgF-Ui). It makes it possible for Calypso to load images from remote private site by using public-api API endpoint instead of requesting the image directly. The changes include:

* Data layer which specifies its `responseType` is a `blob`
* `ProxiedImage` component which calls the above & uses URL.createObjectURL to set the image to the response Blob
* Use `ProxiedImage` when a site is Atomic & Private

#### Testing instructions

* Apply D39236-code (proxy provider script)
* Sandbox the REST API
* Visit `/media` for a Private Atomic site, confirm all images are displayed correctly. Note the loading is going to take some time as we don't support thumbnails yet.
* Visit `/media` for a a regular non-private Atomic site, confirm all images are displayed correctly.

#### Notes

* This PR doesn't address specific images preview/editing just yet

### CODE REVIEW

File: client/my-sites/media-library/proxied-image.jsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import { noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+
+const { Blob } = globalThis; // The linter complains if I don't do this...?
+
+export default function( { alt, mediaUrl, onLoad = noop, siteSlug, style } ) {
```
Comment (adamziel) [inline client/my-sites/media-library/proxied-image.jsx]: Naming this function is the difference between seeing `Anonymous` and `ProxiedImage` in dev tools:

Zrzut ekranu 2020-02-24 o 12 38 41

Zrzut ekranu 2020-02-24 o 12 39 30

```suggestion
export default function ProxiedImage( { alt, mediaUrl, onLoad = noop, siteSlug, style } ) {
```

**Comment by adamziel** [comment]
I added a dead-simple retrying logic

File: client/my-sites/media-library/proxied-image.jsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+			return;
+		}
+
+		wpcom.undocumented().getAtomicSiteMediaViaProxyRetry( siteSlug, mediaUrl, ( err, data ) => {
+			if ( ! ( data instanceof Blob ) ) {
+				setImageData( false );
+				return;
+			}
+
+			const imageUrl = URL.createObjectURL( data );
```
Comment (jblz) [inline client/my-sites/media-library/proxied-image.jsx]: Per: https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL

...we should make sure to call `URL.revokeObjectURL()` on these when the component unmounts or the memory won't be released until the document is unloaded.

File: client/lib/wpcom-undocumented/lib/undocumented.js [hunk 3]
```diff
@@ -2462,4 +2462,25 @@ Undocumented.prototype.startMigration = function( sourceSiteId, targetSiteId ) {
...
+			apiNamespace: 'wpcom/v2',
+			responseType: 'blob',
+		},
+		fn
+	);
+};
+
+Undocumented.prototype.getAtomicSiteMediaViaProxyRetry = function( siteIdOrSlug, mediaPath, fn ) {
+	this.getAtomicSiteMediaViaProxy( siteIdOrSlug, mediaPath, function( err, data ) {
+		if ( err ) {
```
Comment (jblz) [inline client/lib/wpcom-undocumented/lib/undocumented.js]: I don't think this works as-is: when a request fails on the remote site, the response to the browser will still 200 & the `err` condition does not hit.

I'm working on a solution using the `data-layer` code that has built-in retrying, but it's not ready yet.

File: client/my-sites/media-library/proxied-image.jsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		return () => imageData && URL.revokeObjectURL( imageData );
+	} );
+
+	return (
+		<img
+			src={ imageData }
+			onLoad={ onLoad }
+			alt={ alt }
+			style={ style }
+			className="media-library__list-item-centered"
```
Comment (simison) [inline client/my-sites/media-library/proxied-image.jsx]: Nit; should the `className` be kept in `client/my-sites/media-library/list-item-image.jsx` and passed here as an attribute?

**Comment by jblz** [comment]
This is now auto retrying when combined to the change in the proxy provider script.

TODO: investigate & fix the broken tests

**Comment by adamziel** [comment]
Awesome job so far! `<ProxiedImage />` shares a lot of props with regular `` so maybe both these cases should be covered by the same component? Especially since we need to pass the same query params to both.

**Comment by jblz** [comment]
This doesn't work for the "Media Modal"...yet:

* Click an image to select
* Click the `Edit` button
  * See the fake placeholder & a 404 as it tries to fetch the image from the remote site:

Screen Shot 2020-02-28 at 12 41 15 PM
 
We'll need to try to integrate it around here:

https://github.com/Automattic/wp-calypso/blob/559ab78966f370dfe0ed5c701261952d0de63dbe/client/post-editor/media-modal/detail/detail-preview-image.jsx#L87-L103

**Comment by jblz** [comment]
@adamziel -- we can now move between sections & it even uses the data in the cache from the same pageload.  If we move between sections & come back after the "freshness" param expires (currently 30 seconds), it will re-fetch & swap out the images.  Not sure it's necessary, but not a lot of downside to having it that freshness param in place, IMO.

The data does not persist between page loads, but this seems pretty good for a v1 as far as the client cache.

**Comment by adamziel** [comment]
@jblz Amazing job! And yes, we don't need persistent caching for v1. Eventually it would be great to leverage browser caching anyway. This is definitely moving the right way!

File: client/my-sites/media-library/media-image.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	const siteSlug = getSelectedSiteSlug( state ) as string;
+	const isAtomic = !! isSiteAutomatedTransfer( state, siteId as number );
+	const isPrivate = !! isPrivateSite( state, siteId );
+	const { filePath, query, isRelativeToSiteRoot } = parseMediaURL( src, siteSlug );
+
+	const useProxy =
+		// Site privacy / coming soon are getting clobbered when the media lib loads
+		// Should be fixed when D39264-code lands
+		// Hard-coding for now so I can keep working...
+		( true || ( isAtomic && isPrivate ) ) && filePath && isRelativeToSiteRoot;
```
Comment (jblz) [inline client/my-sites/media-library/media-image.tsx]:
```suggestion
		const useProxy = isAtomic && isPrivate && filePath && isRelativeToSiteRoot;
```

D39264-code landed, so this should be functioning reliably now, right?  If not, do we know why yet?

**Review by adamziel — APPROVED** [review]

**Pull request merged**